circuits/spartan/src/eth_membership.rs

```diff
@@ -0,0 +1,188 @@
+use ark_ff::PrimeField;
+use sapir::{
+    constraint_system::{ConstraintSystem, Wire},
+    frontend::gadgets::{
+        ec_add_complete, ec_mul, poseidon::poseidon::PoseidonChip, to_addr, to_le_bits,
+        verify_merkle_proof, AffinePoint,
+    },
+    poseidon::constants::secp256k1_w3,
+};
+
+pub const TREE_DEPTH: usize = 15;
+pub const NUM_MERKLE_PROOFS: usize = 1;
+
+pub struct AssignedMerkleProof<F: PrimeField> {
+    pub siblings: Vec<Wire<F>>,
+    pub indices: Vec<Wire<F>>,
+}
+
+impl<F: PrimeField> AssignedMerkleProof<F> {
+    pub fn new(siblings: Vec<Wire<F>>, indices: Vec<Wire<F>>) -> Self {
+        Self { siblings, indices }
+    }
+}
+
+pub fn eth_membership<F: PrimeField>(cs: &mut ConstraintSystem<F>) {
+    // #############################################
+    // Private inputs
+    // #############################################
+
+    // `s` part of the signature
+    let s_bits = cs.alloc_priv_inputs(256);
+
+    let mut merkle_proofs = Vec::with_capacity(NUM_MERKLE_PROOFS);
+
+    for _ in 0..NUM_MERKLE_PROOFS {
+        let merkle_indices = cs.alloc_priv_inputs(TREE_DEPTH);
+        let merkle_siblings = cs.alloc_priv_inputs(TREE_DEPTH);
+
+        merkle_proofs.push(AssignedMerkleProof::new(merkle_siblings, merkle_indices));
+    }
+
+    // #############################################
+    // Public inputs
+    // #############################################
+
+    let t_x = cs.alloc_pub_input();
+    let t_y = cs.alloc_pub_input();
+
+    let u_x = cs.alloc_pub_input();
+    let u_y = cs.alloc_pub_input();
+
+    // #############################################
+    // Constraints
+    // #############################################
+
+    // 1. Recover the public key from the signature
+    // s_mul_t = s * T
+    let t = AffinePoint::new(t_x, t_y);
+    let s_mul_t = ec_mul(t, &s_bits, cs);
+
+    // pubKey = sMultT + U
+    let u = AffinePoint::new(u_x, u_y);
+    let pub_key = ec_add_complete(s_mul_t, u, cs);
+
+    let pub_key_x_bits = to_le_bits(pub_key.x);
+    let pub_key_y_bits = to_le_bits(pub_key.y);
+
+    // We need this transformation because the bits should be in little endian
+    // and the bytes should be in big endian.
```

Comment:
nit (I had to read this 3 times to see there was a difference :rofl: )
```suggestion
    // We need this transformation because the BITS should be in LITTLE endian
    // and the BYTES should be in BIG endian.
```